In [1]:
import pandas as pd

#import statsmodels.formula.api as smf
#import statsmodels.api as sm

from sklearn.linear_model import LogisticRegression

import numpy as np

from sklearn.discriminant_analysis import LinearDiscriminantAnalysis, QuadraticDiscriminantAnalysis
from sklearn.metrics import confusion_matrix, classification_report

from sklearn.metrics import accuracy_score

from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB

from sklearn.model_selection import train_test_split

from sklearn import neighbors

import warnings
warnings.simplefilter("ignore")

# 1. Data Preparation

## 1.1 Read and Clean Data

    Perform following actions on AimoScore and WeakLinks
    Read,
    Print Lenght
    Remove ? and NAN
    Print Lenght
    Print Summary

In [2]:
url = "../AimoScore_WeakLink_big_scores.xls"
AimoScore = pd.read_excel(url)
print(len(AimoScore))
data = AimoScore[~AimoScore.isin(['?'])]
data = data.dropna(axis=0)
AimoScore = pd.DataFrame(data)
print(len(AimoScore))
AimoScore.describe()

2094
2094


AimoScore  No_1_Angle_Deviation  No_2_Angle_Deviation  \
count  2094.000000           2094.000000           2094.000000   
mean      0.572740              0.539248              0.504462   
std       0.236510              0.239284              0.282136   
min       0.010000              0.279770              0.092300   
25%       0.407907              0.279770              0.250149   
50%       0.617248              0.500000              0.500000   
75%       0.759378              0.749761              0.750239   
max       0.993987              1.000000              1.000000   

       No_3_Angle_Deviation  No_4_Angle_Deviation  No_5_Angle_Deviation  \
count           2094.000000           2094.000000           2094.000000   
mean               0.560258              0.518065              0.513932   
std                0.217828              0.264104              0.270076   
min                0.346724              0.188905              0.163080   
25%                0.346724              0.250239              0.250717   
50%                0.500000              0.500000              0.500478   
75%                0.750239              0.750239              0.750239   
max                1.000000              1.000000              1.000000   

       No_6_Angle_Deviation  No_7_Angle_Deviation  No_8_Angle_Deviation  \
count           2094.000000           2094.000000           2094.000000   
mean               0.510336              0.519434              0.635923   
std                0.274473              0.262884              0.153280   
min                0.140124              0.194165              0.303247   
25%                0.250717              0.250717              0.521760   
50%                0.500478              0.500956              0.521760   
75%                0.750239              0.750239              0.749761   
max                1.000000              1.000000              1.000000   

       No_9_Angle_Deviation  ...  No_19_NASM_Deviation  No_20_NASM_Deviation  \
count           2094.000000  ...           2094.000000           2094.000000   
mean               0.512548  ...              0.725505              0.715943   
std                0.271193  ...              0.095008              0.100437   
min                0.156385  ...              0.670971              0.656624   
25%                0.250717  ...              0.670971              0.656624   
50%                0.500478  ...              0.670971              0.656624   
75%                0.750239  ...              0.751196              0.751196   
max                1.000000  ...              1.000000              1.000000   

       No_21_NASM_Deviation  No_22_NASM_Deviation  No_23_NASM_Deviation  \
count           2094.000000           2094.000000           2094.000000   
mean               0.706319              0.652969              0.710774   
std                0.105886              0.141000              0.103196   
min                0.640401              0.525310              0.648972   
25%                0.642276              0.552846              0.648972   
50%                0.642276              0.552846              0.648972   
75%                0.749761              0.749880              0.751196   
max                1.000000              1.000000              1.000000   

       No_24_NASM_Deviation  No_25_NASM_Deviation  No_1_Time_Deviation  \
count           2094.000000           2094.000000          2094.000000   
mean               0.667517              0.548089             0.512021   
std                0.130989              0.230642             0.273056   
min                0.578192              0.308943             0.148733   
25%                0.578192              0.308943             0.250717   
50%                0.578192              0.500000             0.500478   
75%                0.751196              0.750239             0.750239   
max                1.000000              1.000000             1.000000   

In [3]:
url = "../20190108 scores_and_weak_links.xlsx"
WeakLinks = pd.read_excel(url, engine='openpyxl')
print(len(WeakLinks))
data = WeakLinks[~WeakLinks.isin(['?'])]
data = data.dropna(axis=0)
WeakLinks = pd.DataFrame(data)
print(len(WeakLinks))
WeakLinks.describe()

2096
2096


SCORE  ForwardHead  LeftArmFallForward  RightArmFallForward  \
count  2096.000000  2096.000000         2096.000000          2096.000000   
mean      0.572420     0.301579            0.996842             0.880792   
std       0.236856     0.477732            0.481902             0.578039   
min       0.010000     0.000000            0.000000             0.000000   
25%       0.407843     0.000000            0.690000             0.170000   
50%       0.617248     0.000000            1.120000             1.110000   
75%       0.759416     0.480000            1.390000             1.420000   
max       0.993987     1.500000            1.500000             1.500000   

       LeftShoulderElevation  RightShoulderElevation  ExcessiveForwardLean  \
count            2096.000000             2096.000000           2096.000000   
mean                0.271293                0.465000              0.201150   
std                 0.418118                0.475603              0.419338   
min                 0.000000                0.000000              0.000000   
25%                 0.000000                0.030000              0.000000   
50%                 0.030000                0.325000              0.000000   
75%                 0.390000                0.852500              0.080000   
max                 1.500000                1.500000              1.500000   

       LeftAsymmetricalWeightShift  RightAsymmetricalWeightShift  \
count                  2096.000000                   2096.000000   
mean                      0.175219                      0.065315   
std                       0.397579                      0.252260   
min                       0.000000                      0.000000   
25%                       0.000000                      0.000000   
50%                       0.000000                      0.000000   
75%                       0.050000                      0.000000   
max                       1.500000                      1.500000   

       LeftKneeMovesInward  RightKneeMovesInward  LeftKneeMovesOutward  \
count          2096.000000           2096.000000           2096.000000   
mean              0.037657              0.179089              0.170787   
std               0.140765              0.329131              0.401551   
min               0.000000              0.000000              0.000000   
25%               0.000000              0.000000              0.000000   
50%               0.000000              0.000000              0.000000   
75%               0.000000              0.200000              0.020000   
max               1.310000              1.500000              1.500000   

       RightKneeMovesOutward  LeftHeelRises  RightHeelRises  
count            2096.000000    2096.000000     2096.000000  
mean                0.238912       0.059241        0.102505  
std                 0.528886       0.226736        0.290971  
min                 0.000000       0.000000        0.000000  
25%                 0.000000       0.000000        0.000000  
50%                 0.000000       0.000000        0.000000  
75%                 0.000000       0.000000        0.000000  
max                 1.500000       1.500000        1.500000

### <font color='green'>Data was clean. not found any ? and NAN</font>

# 2 Analysis on IDs

## 2.1 Count IDs

In [4]:
print(len(AimoScore['ID'].value_counts()))
print(len(WeakLinks['ID'].value_counts()))

2091
2096


### <font color='red'>AimoScore has 2094 values but 2091 IDs. which mean 3 rows either have duplicate IDs OR empty</font>
### <font color='green'>WeakLinks is OK</font>

## 2.2 Cross verify IDs
    
    Check which ID of AimoScore are in WeakLinks 
    Check which ID of WeakLinks are in AimoScore

In [5]:
WeakLinks['ID'].isin(AimoScore['ID']).value_counts()

True     2091
False       5
Name: ID, dtype: int64

In [6]:
AimoScore['ID'].isin(WeakLinks['ID']).value_counts()

True    2094
Name: ID, dtype: int64

### <font color='red'>AimoScore has 2094 values matched to WeakLinks. which mean either WeakLinks have empty ID or some IDs in AimoScore are duplicate</font>
### <font color='red'>WeakLinks has 2091 values matched to AimoScore. which mean 5 extra IDs</font>

## 2.3 Find differnt IDs
    
    Check differnet IDs using symmetric_difference and concat

In [7]:
set(AimoScore.ID).symmetric_difference(WeakLinks.ID)

{'076c6772-6b42-4b95-bb8a-8a92cf523177.Kinect',
 '18dbbf0b-2828-40ba-89ab-d8cd53f780d0.Kinect',
 '1e6e9935-5aff-4d1a-8410-f45b255d98c8.Kinect',
 '49a562eb-07a9-4cda-8d50-ab85b6c50d0a.Kinect',
 'ab5aa0a5-1e82-4d9b-8bd4-5a12cfc92c92.Kinect'}

In [8]:
combined_dfs = pd.concat([AimoScore['ID'], WeakLinks['ID']])
combined_dfs.drop_duplicates(keep=False)

277     076c6772-6b42-4b95-bb8a-8a92cf523177.Kinect
809     ab5aa0a5-1e82-4d9b-8bd4-5a12cfc92c92.Kinect
953     1e6e9935-5aff-4d1a-8410-f45b255d98c8.Kinect
1149    18dbbf0b-2828-40ba-89ab-d8cd53f780d0.Kinect
1201    49a562eb-07a9-4cda-8d50-ab85b6c50d0a.Kinect
Name: ID, dtype: object

### <font color='green'>We found 5 IDs that differ</font>

## 2.4 Check Existense of IDs
    Using indexes, and merge, check IDs exist in One OR in both DataFrames

In [9]:
print(AimoScore[AimoScore['ID'] == "076c6772-6b42-4b95-bb8a-8a92cf523177.Kinect"].index)
print(AimoScore[AimoScore['ID'] == "ab5aa0a5-1e82-4d9b-8bd4-5a12cfc92c92.Kinect"].index)
print(AimoScore[AimoScore['ID'] == "1e6e9935-5aff-4d1a-8410-f45b255d98c8.Kinect"].index)
print(AimoScore[AimoScore['ID'] == "18dbbf0b-2828-40ba-89ab-d8cd53f780d0.Kinect"].index)
print(AimoScore[AimoScore['ID'] == "49a562eb-07a9-4cda-8d50-ab85b6c50d0a.Kinect"].index)
print(WeakLinks[WeakLinks['ID'] == "076c6772-6b42-4b95-bb8a-8a92cf523177.Kinect"].index)
print(WeakLinks[WeakLinks['ID'] == "ab5aa0a5-1e82-4d9b-8bd4-5a12cfc92c92.Kinect"].index)
print(WeakLinks[WeakLinks['ID'] == "1e6e9935-5aff-4d1a-8410-f45b255d98c8.Kinect"].index)
print(WeakLinks[WeakLinks['ID'] == "18dbbf0b-2828-40ba-89ab-d8cd53f780d0.Kinect"].index)
print(WeakLinks[WeakLinks['ID'] == "49a562eb-07a9-4cda-8d50-ab85b6c50d0a.Kinect"].index)

Int64Index([], dtype='int64')
Int64Index([], dtype='int64')
Int64Index([], dtype='int64')
Int64Index([], dtype='int64')
Int64Index([], dtype='int64')
Int64Index([277], dtype='int64')
Int64Index([809], dtype='int64')
Int64Index([953], dtype='int64')
Int64Index([1149], dtype='int64')
Int64Index([1201], dtype='int64')


In [10]:
AimoScore.merge(WeakLinks, how = 'outer' ,indicator=True).loc[lambda x : x['_merge']=='left_only']

Empty DataFrame
Columns: [AimoScore, No_1_Angle_Deviation, No_2_Angle_Deviation, No_3_Angle_Deviation, No_4_Angle_Deviation, No_5_Angle_Deviation, No_6_Angle_Deviation, No_7_Angle_Deviation, No_8_Angle_Deviation, No_9_Angle_Deviation, No_10_Angle_Deviation, No_11_Angle_Deviation, No_12_Angle_Deviation, No_13_Angle_Deviation, No_1_NASM_Deviation, No_2_NASM_Deviation, No_3_NASM_Deviation, No_4_NASM_Deviation, No_5_NASM_Deviation, No_6_NASM_Deviation, No_7_NASM_Deviation, No_8_NASM_Deviation, No_9_NASM_Deviation, No_10_NASM_Deviation, No_11_NASM_Deviation, No_12_NASM_Deviation, No_13_NASM_Deviation, No_14_NASM_Deviation, No_15_NASM_Deviation, No_16_NASM_Deviation, No_17_NASM_Deviation, No_18_NASM_Deviation, No_19_NASM_Deviation, No_20_NASM_Deviation, No_21_NASM_Deviation, No_22_NASM_Deviation, No_23_NASM_Deviation, No_24_NASM_Deviation, No_25_NASM_Deviation, No_1_Time_Deviation, No_2_Time_Deviation, EstimatedScore, ID, Date, SCORE, ForwardHead, LeftArmFallForward, RightArmFallForward, LeftShoulderElevation, RightShoulderElevation, ExcessiveForwardLean, LeftAsymmetricalWeightShift, RightAsymmetricalWeightShift, LeftKneeMovesInward, RightKneeMovesInward, LeftKneeMovesOutward, RightKneeMovesOutward, LeftHeelRises, RightHeelRises, _merge]
Index: []

[0 rows x 60 columns]

In [11]:
AimoScore.merge(WeakLinks, how = 'outer' ,indicator=True).loc[lambda x : x['_merge']=='right_only']

AimoScore  No_1_Angle_Deviation  No_2_Angle_Deviation  \
2094        NaN                   NaN                   NaN   
2095        NaN                   NaN                   NaN   
2096        NaN                   NaN                   NaN   
2097        NaN                   NaN                   NaN   
2098        NaN                   NaN                   NaN   

      No_3_Angle_Deviation  No_4_Angle_Deviation  No_5_Angle_Deviation  \
2094                   NaN                   NaN                   NaN   
2095                   NaN                   NaN                   NaN   
2096                   NaN                   NaN                   NaN   
2097                   NaN                   NaN                   NaN   
2098                   NaN                   NaN                   NaN   

      No_6_Angle_Deviation  No_7_Angle_Deviation  No_8_Angle_Deviation  \
2094                   NaN                   NaN                   NaN   
2095                   NaN                   NaN                   NaN   
2096                   NaN                   NaN                   NaN   
2097                   NaN                   NaN                   NaN   
2098                   NaN                   NaN                   NaN   

      No_9_Angle_Deviation  ...  ExcessiveForwardLean  \
2094                   NaN  ...                  0.00   
2095                   NaN  ...                  0.00   
2096                   NaN  ...                  0.00   
2097                   NaN  ...                  0.00   
2098                   NaN  ...                  1.39   

      LeftAsymmetricalWeightShift  RightAsymmetricalWeightShift  \
2094                          0.0                           0.0   
2095                          0.0                           0.0   
2096                          0.0                           0.0   
2097                          0.0                           0.0   
2098                          0.0                           0.0   

      LeftKneeMovesInward  RightKneeMovesInward  LeftKneeMovesOutward  \
2094                  0.0                  0.00                  0.00   
2095                  0.0                  0.00                  0.02   
2096                  0.0                  0.00                  0.02   
2097                  0.0                  0.00                  0.00   
2098                  0.0                  0.01                  0.00   

      RightKneeMovesOutward  LeftHeelRises  RightHeelRises      _merge  
2094                    0.0           0.00            0.00  right_only  
2095                    1.5           0.35            1.04  right_only  
2096                    1.5           0.68            1.14  right_only  
2097                    0.0           0.00            0.00  right_only  
2098                    0.0           0.00            0.00  right_only  

[5 rows x 60 columns]

### <font color='red'>IDs not found in AimoScore</font>
### <font color='green'>WeakLinks has all 5 extra IDs</font>
### <font color='red'>We can conclude that AimoScore has 3 duplicate IDs</font>

## 2.5 Find duplicate ID record
    Check all record that contain duplicate IDs

In [12]:
AimoScore.groupby("ID").filter(lambda x: len(x) > 1)

AimoScore  No_1_Angle_Deviation  No_2_Angle_Deviation  \
2088   0.527640              0.575323              0.193687   
2089   0.407197              0.279770              0.131038   
2090   0.965930              0.564802              0.311813   
2091   0.527640              0.404489              0.214422   
2092   0.407197              0.606972              0.863897   
2093   0.965930              0.479943              0.250239   

      No_3_Angle_Deviation  No_4_Angle_Deviation  No_5_Angle_Deviation  \
2088              0.780488              0.956958              0.740794   
2089              0.346724              0.317073              0.945481   
2090              0.376375              0.188905              0.187948   
2091              0.822827              0.937440              0.739255   
2092              0.467526              0.225406              0.946036   
2093              0.366762              0.316141              0.489494   

      No_6_Angle_Deviation  No_7_Angle_Deviation  No_8_Angle_Deviation  \
2088              0.912482              0.630799              0.632233   
2089              0.646581              0.913439              0.521760   
2090              0.347202              0.227642              0.521760   
2091              0.909265              0.681471              0.658071   
2092              0.706781              0.934097              0.303247   
2093              0.444126              0.681471              0.303247   

      No_9_Angle_Deviation  ...  No_20_NASM_Deviation  No_21_NASM_Deviation  \
2088              0.708274  ...              0.656624              0.642276   
2089              0.570540  ...              0.945002              0.642276   
2090              0.248685  ...              0.753228              0.642276   
2091              0.822350  ...              0.886819              0.640401   
2092              0.576409  ...              0.886819              0.640401   
2093              0.266953  ...              0.886819              0.640401   

      No_22_NASM_Deviation  No_23_NASM_Deviation  No_24_NASM_Deviation  \
2088              0.552846              0.648972              0.578192   
2089              0.596365              0.648972              0.749880   
2090              0.552846              0.648972              0.869440   
2091              0.525310              0.776982              0.822827   
2092              0.725883              0.757402              0.822827   
2093              0.525310              0.757402              0.877746   

      No_25_NASM_Deviation  No_1_Time_Deviation  No_2_Time_Deviation  \
2088              0.736011             0.764706             0.811095   
2089              0.943568             0.684840             0.711621   
2090              0.316117             0.148733             0.151124   
2091              0.367717             0.584050             0.570201   
2092              0.943649             0.953200             0.956543   
2093              0.444604             0.419293             0.423591   

      EstimatedScore                                           ID  
2088        0.911047  ff015ddd-c9fb-49e3-8e33-abbf72e8b120.Kinect  
2089        0.877571  ff514cba-6c5e-41a9-9936-3e8400565c8d.Kinect  
2090        0.030129  ff63b8a9-deb5-4179-b424-fd07d04ca37f.Kinect  
2091        0.613658  ff015ddd-c9fb-49e3-8e33-abbf72e8b120.Kinect  
2092        0.761700  ff514cba-6c5e-41a9-9936-3e8400565c8d.Kinect  
2093        0.042025  ff63b8a9-deb5-4179-b424-fd07d04ca37f.Kinect  

[6 rows x 43 columns]

In [13]:
WeakLinks.groupby("ID").filter(lambda x: len(x) > 1)

Empty DataFrame
Columns: [ID, Date, SCORE, ForwardHead, LeftArmFallForward, RightArmFallForward, LeftShoulderElevation, RightShoulderElevation, ExcessiveForwardLean, LeftAsymmetricalWeightShift, RightAsymmetricalWeightShift, LeftKneeMovesInward, RightKneeMovesInward, LeftKneeMovesOutward, RightKneeMovesOutward, LeftHeelRises, RightHeelRises]
Index: []

### <font color='green'>We found duplicate ID records in AimoScore</font>

# <font color='green'>Result</font>

    AimoScore has 2094 records with 2091 unique IDs, and 3 Repeating IDs
        - ff015ddd-c9fb-49e3-8e33-abbf72e8b120.Kinect at position 2088, 2091
        - ff514cba-6c5e-41a9-9936-3e8400565c8d.Kinect at postion 2089, 2092	
        - ff63b8a9-deb5-4179-b424-fd07d04ca37f.Kinect at postion 2090, 2093
    
    WeakLinks has 2096 records with 2096 unique IDs
    
    AimoScore's all 2091 (remember 3 IDs areduplicate) IDs have matching IDs in WeakLinks
     
    WeakLinks has 5 extra IDs
        - 076c6772-6b42-4b95-bb8a-8a92cf523177.Kinect at position 277
        - ab5aa0a5-1e82-4d9b-8bd4-5a12cfc92c92.Kinect at position 809     
        - 1e6e9935-5aff-4d1a-8410-f45b255d98c8.Kinect at position 953
        - 18dbbf0b-2828-40ba-89ab-d8cd53f780d0.Kinect at position 1149
        - 49a562eb-07a9-4cda-8d50-ab85b6c50d0a.Kinect at position 1201
        
    AimoScore's duplicate IDs data/record (featrues) values are differnet (Only IDs are repeating). and we are not sure that all IDs need to be unique OR we don't have any other crietria to drop one of duplicatie IDs record. so we will keep these and will find WeakLink's as it is.
        
        


# 3 Join DataSet

## 3.1 Join DataSet
    Join DataSet using ID
    Check for any NAN values

In [14]:
AimoScoreWeakLinks = pd.merge(AimoScore, WeakLinks,on=["ID"])
AimoScoreWeakLinks.describe()

AimoScore  No_1_Angle_Deviation  No_2_Angle_Deviation  \
count  2094.000000           2094.000000           2094.000000   
mean      0.572740              0.539248              0.504462   
std       0.236510              0.239284              0.282136   
min       0.010000              0.279770              0.092300   
25%       0.407907              0.279770              0.250149   
50%       0.617248              0.500000              0.500000   
75%       0.759378              0.749761              0.750239   
max       0.993987              1.000000              1.000000   

       No_3_Angle_Deviation  No_4_Angle_Deviation  No_5_Angle_Deviation  \
count           2094.000000           2094.000000           2094.000000   
mean               0.560258              0.518065              0.513932   
std                0.217828              0.264104              0.270076   
min                0.346724              0.188905              0.163080   
25%                0.346724              0.250239              0.250717   
50%                0.500000              0.500000              0.500478   
75%                0.750239              0.750239              0.750239   
max                1.000000              1.000000              1.000000   

       No_6_Angle_Deviation  No_7_Angle_Deviation  No_8_Angle_Deviation  \
count           2094.000000           2094.000000           2094.000000   
mean               0.510336              0.519434              0.635923   
std                0.274473              0.262884              0.153280   
min                0.140124              0.194165              0.303247   
25%                0.250717              0.250717              0.521760   
50%                0.500478              0.500956              0.521760   
75%                0.750239              0.750239              0.749761   
max                1.000000              1.000000              1.000000   

       No_9_Angle_Deviation  ...  RightShoulderElevation  \
count           2094.000000  ...             2094.000000   
mean               0.512548  ...                0.465277   
std                0.271193  ...                0.475690   
min                0.156385  ...                0.000000   
25%                0.250717  ...                0.030000   
50%                0.500478  ...                0.325000   
75%                0.750239  ...                0.857500   
max                1.000000  ...                1.500000   

       ExcessiveForwardLean  LeftAsymmetricalWeightShift  \
count           2094.000000                  2094.000000   
mean               0.201098                     0.175435   
std                0.418949                     0.397713   
min                0.000000                     0.000000   
25%                0.000000                     0.000000   
50%                0.000000                     0.000000   
75%                0.080000                     0.050000   
max                1.500000                     1.500000   

       RightAsymmetricalWeightShift  LeftKneeMovesInward  \
count                   2094.000000          2094.000000   
mean                       0.065377             0.037693   
std                        0.252373             0.140828   
min                        0.000000             0.000000   
25%                        0.000000             0.000000   
50%                        0.000000             0.000000   
75%                        0.000000             0.000000   
max                        1.500000             1.310000   

       RightKneeMovesInward  LeftKneeMovesOutward  RightKneeMovesOutward  \
count           2094.000000           2094.000000            2094.000000   
mean               0.179255              0.170941               0.238424   
std                0.329244              0.401712               0.528395   
min                0.000000              0.000000               0.000000   
25%                0.000000              0.000000          

In [15]:
AimoScoreWeakLinks[AimoScoreWeakLinks.isna().any(axis=1)]

Empty DataFrame
Columns: [AimoScore, No_1_Angle_Deviation, No_2_Angle_Deviation, No_3_Angle_Deviation, No_4_Angle_Deviation, No_5_Angle_Deviation, No_6_Angle_Deviation, No_7_Angle_Deviation, No_8_Angle_Deviation, No_9_Angle_Deviation, No_10_Angle_Deviation, No_11_Angle_Deviation, No_12_Angle_Deviation, No_13_Angle_Deviation, No_1_NASM_Deviation, No_2_NASM_Deviation, No_3_NASM_Deviation, No_4_NASM_Deviation, No_5_NASM_Deviation, No_6_NASM_Deviation, No_7_NASM_Deviation, No_8_NASM_Deviation, No_9_NASM_Deviation, No_10_NASM_Deviation, No_11_NASM_Deviation, No_12_NASM_Deviation, No_13_NASM_Deviation, No_14_NASM_Deviation, No_15_NASM_Deviation, No_16_NASM_Deviation, No_17_NASM_Deviation, No_18_NASM_Deviation, No_19_NASM_Deviation, No_20_NASM_Deviation, No_21_NASM_Deviation, No_22_NASM_Deviation, No_23_NASM_Deviation, No_24_NASM_Deviation, No_25_NASM_Deviation, No_1_Time_Deviation, No_2_Time_Deviation, EstimatedScore, ID, Date, SCORE, ForwardHead, LeftArmFallForward, RightArmFallForward, LeftShoulderElevation, RightShoulderElevation, ExcessiveForwardLean, LeftAsymmetricalWeightShift, RightAsymmetricalWeightShift, LeftKneeMovesInward, RightKneeMovesInward, LeftKneeMovesOutward, RightKneeMovesOutward, LeftHeelRises, RightHeelRises]
Index: []

[0 rows x 59 columns]

## 3.2 Find Weakest link
    Finding the weak link using weak link columns(14) that has maximum score in each row and adding it as a new column 'WeakLinks'

In [16]:
AimoScoreWeakLinks["WeakLinks"] = AimoScoreWeakLinks.loc[:,"ForwardHead":"RightHeelRises"].idxmax(axis=1)

In [17]:
cols = AimoScoreWeakLinks.columns
quantative_cols = AimoScoreWeakLinks._get_numeric_data().columns
qualitative_cols = list(set(cols) - set(quantative_cols))

print('\nquantative predictors/variables')
for index, col in enumerate(quantative_cols, start=1):
   print (index, '.\'', col, '\' ', sep='', end='')

print('\n\nqualitative predictors/variables')
for index, col in enumerate(qualitative_cols, start=1):
   print (index, '.\'', col, '\' ', sep='', end='')


quantative predictors/variables
1.'AimoScore' 2.'No_1_Angle_Deviation' 3.'No_2_Angle_Deviation' 4.'No_3_Angle_Deviation' 5.'No_4_Angle_Deviation' 6.'No_5_Angle_Deviation' 7.'No_6_Angle_Deviation' 8.'No_7_Angle_Deviation' 9.'No_8_Angle_Deviation' 10.'No_9_Angle_Deviation' 11.'No_10_Angle_Deviation' 12.'No_11_Angle_Deviation' 13.'No_12_Angle_Deviation' 14.'No_13_Angle_Deviation' 15.'No_1_NASM_Deviation' 16.'No_2_NASM_Deviation' 17.'No_3_NASM_Deviation' 18.'No_4_NASM_Deviation' 19.'No_5_NASM_Deviation' 20.'No_6_NASM_Deviation' 21.'No_7_NASM_Deviation' 22.'No_8_NASM_Deviation' 23.'No_9_NASM_Deviation' 24.'No_10_NASM_Deviation' 25.'No_11_NASM_Deviation' 26.'No_12_NASM_Deviation' 27.'No_13_NASM_Deviation' 28.'No_14_NASM_Deviation' 29.'No_15_NASM_Deviation' 30.'No_16_NASM_Deviation' 31.'No_17_NASM_Deviation' 32.'No_18_NASM_Deviation' 33.'No_19_NASM_Deviation' 34.'No_20_NASM_Deviation' 35.'No_21_NASM_Deviation' 36.'No_22_NASM_Deviation' 37.'No_23_NASM_Deviation' 38.'No_24_NASM_Deviation' 39.'

In [18]:
AimoScoreWeakLinks["WeakLinks"].value_counts()

LeftArmFallForward              630
RightArmFallForward             472
RightKneeMovesOutward           277
RightShoulderElevation          250
ForwardHead                     115
ExcessiveForwardLean            100
LeftAsymmetricalWeightShift      73
LeftShoulderElevation            58
LeftKneeMovesOutward             49
RightKneeMovesInward             45
RightAsymmetricalWeightShift     19
LeftKneeMovesInward               3
LeftHeelRises                     2
RightHeelRises                    1
Name: WeakLinks, dtype: int64

In [19]:
temp = (AimoScoreWeakLinks["WeakLinks"].value_counts()<2)
temp = temp[~temp == False]
AimoScoreWeakLinks = AimoScoreWeakLinks[~AimoScoreWeakLinks["WeakLinks"].isin(temp.index)]
AimoScoreWeakLinks["WeakLinks"].value_counts()

LeftArmFallForward              630
RightArmFallForward             472
RightKneeMovesOutward           277
RightShoulderElevation          250
ForwardHead                     115
ExcessiveForwardLean            100
LeftAsymmetricalWeightShift      73
LeftShoulderElevation            58
LeftKneeMovesOutward             49
RightKneeMovesInward             45
RightAsymmetricalWeightShift     19
LeftKneeMovesInward               3
LeftHeelRises                     2
Name: WeakLinks, dtype: int64

## 3.3 Discard extra columns
    Discard the 14 weak link scores from the joint table

In [20]:
AimoScoreWeakLinks.drop(AimoScoreWeakLinks.columns.to_series()["ForwardHead":"RightHeelRises"], axis=1, inplace=True)
len(AimoScoreWeakLinks.columns)

46

In [21]:
AimoScoreWeakLinks.describe()

AimoScore  No_1_Angle_Deviation  No_2_Angle_Deviation  \
count  2093.000000           2093.000000           2093.000000   
mean      0.572735              0.539250              0.504328   
std       0.236566              0.239342              0.282137   
min       0.010000              0.279770              0.092300   
25%       0.407847              0.279770              0.250120   
50%       0.617345              0.499761              0.499761   
75%       0.759383              0.749880              0.749880   
max       0.993987              1.000000              1.000000   

       No_3_Angle_Deviation  No_4_Angle_Deviation  No_5_Angle_Deviation  \
count           2093.000000           2093.000000           2093.000000   
mean               0.560361              0.518222              0.514099   
std                0.217830              0.264069              0.270031   
min                0.346724              0.188905              0.163080   
25%                0.346724              0.250598              0.251076   
50%                0.500239              0.500239              0.500717   
75%                0.750359              0.750359              0.750359   
max                1.000000              1.000000              1.000000   

       No_6_Angle_Deviation  No_7_Angle_Deviation  No_8_Angle_Deviation  \
count           2093.000000           2093.000000           2093.000000   
mean               0.510498              0.519589              0.635978   
std                0.274438              0.262851              0.153296   
min                0.140124              0.194165              0.303247   
25%                0.251076              0.251076              0.521760   
50%                0.500717              0.501196              0.521760   
75%                0.750359              0.750359              0.749880   
max                1.000000              1.000000              1.000000   

       No_9_Angle_Deviation  ...  No_20_NASM_Deviation  No_21_NASM_Deviation  \
count           2093.000000  ...           2093.000000           2093.000000   
mean               0.512710  ...              0.715972              0.706350   
std                0.271156  ...              0.100453              0.105902   
min                0.156385  ...              0.656624              0.640401   
25%                0.251076  ...              0.656624              0.642276   
50%                0.500717  ...              0.656624              0.642276   
75%                0.750359  ...              0.751315              0.749880   
max                1.000000  ...              1.000000              1.000000   

       No_22_NASM_Deviation  No_23_NASM_Deviation  No_24_NASM_Deviation  \
count           2093.000000           2093.000000           2093.000000   
mean               0.653017              0.710804              0.667559   
std                0.141017              0.103212              0.131006   
min                0.525310              0.648972              0.578192   
25%                0.552846              0.648972              0.578192   
50%                0.552846              0.648972              0.578192   
75%                0.749880              0.751315              0.751315   
max                1.000000              1.000000              1.000000   

       No_25_NASM_Deviation  No_1_Time_Deviation  No_2_Time_Deviation  \
count           2093.000000          2093.000000          2093.000000   
mean               0.548203             0.512151             0.512546   
std                0.230638             0.273058             0.272598   
min                0.308943             0.148733             0.151124   
25%                0.308943             0.251076             0.251076   
50%                0.500239             0.500717             0.500717   
75%                0.750359             0.750359             0.750359   
max                1.000000             1.000000             1.000000   

       

In [22]:
AimoScoreWeakLinks

AimoScore  No_1_Angle_Deviation  No_2_Angle_Deviation  \
0      0.323667              0.538020              0.815878   
1      0.323699              0.443807              0.306552   
2      0.848327              0.603539              0.373984   
3      0.351332              0.484935              0.623625   
4      0.627181              0.860832              0.657580   
...         ...                   ...                   ...   
2089   0.527640              0.404489              0.214422   
2090   0.407197              0.279770              0.131038   
2091   0.407197              0.606972              0.863897   
2092   0.965930              0.564802              0.311813   
2093   0.965930              0.479943              0.250239   

      No_3_Angle_Deviation  No_4_Angle_Deviation  No_5_Angle_Deviation  \
0                 0.346724              0.382114              0.302248   
1                 0.823529              0.188905              0.497370   
2                 0.346724              0.590626              0.341942   
3                 0.380201              0.975132              0.509326   
4                 0.745576              0.552846              0.375897   
...                    ...                   ...                   ...   
2089              0.822827              0.937440              0.739255   
2090              0.346724              0.317073              0.945481   
2091              0.467526              0.225406              0.946036   
2092              0.376375              0.188905              0.187948   
2093              0.366762              0.316141              0.489494   

      No_6_Angle_Deviation  No_7_Angle_Deviation  No_8_Angle_Deviation  \
0                 0.947872              0.275945              0.521760   
1                 0.140124              0.664275              0.521760   
2                 0.298900              0.276901              0.623625   
3                 0.888570              0.363462              0.847441   
4                 0.483022              0.388331              0.521760   
...                    ...                   ...                   ...   
2089              0.909265              0.681471              0.658071   
2090              0.646581              0.913439              0.521760   
2091              0.706781              0.934097              0.303247   
2092              0.347202              0.227642              0.521760   
2093              0.444126              0.681471              0.303247   

      No_9_Angle_Deviation  ...  No_23_NASM_Deviation  No_24_NASM_Deviation  \
0                 0.457198  ...              0.648972              0.578192   
1                 0.729316  ...              0.848876              0.889048   
2                 0.658058  ...              0.648972              0.578192   
3                 0.237207  ...              0.648972              0.578192   
4                 0.387853  ...              0.648972              0.578192   
...                    ...  ...                   ...                   ...   
2089              0.822350  ...              0.776982              0.822827   
2090              0.570540  ...              0.648972              0.749880   
2091              0.576409  ...              0.757402              0.822827   
2092              0.248685  ...              0.648972              0.869440   
2093              0.266953  ...              0.757402              0.877746   

      No_25_NASM_Deviation  No_1_Time_Deviation  No_2_Time_Deviation  \
0                 0.560019             0.821616             0.818747   
1                 0.816834             0.307987             0.248207   
2                 0.555715             0.218556             0.235294   
3                 0.744620             0.458154             0.432807   
4                 0.308943             0.805356             0.774271   
...                    ...                  ...                  ...   
2089              

## 4 Classification

### <font color='green'>38 Predictors</font>

Use 80% with 38 predictors (all angle deviations and nasm deviations)

In [23]:
trainingPercent = 0.8

y_train, y_test ,X_train, X_test  = train_test_split(AimoScoreWeakLinks[["WeakLinks"]], AimoScoreWeakLinks.drop(columns=["WeakLinks","AimoScore","SCORE","ID","Date","No_1_Time_Deviation","No_2_Time_Deviation","EstimatedScore"]), train_size = trainingPercent)

### Logestic Regression

In [24]:
lr_ovr = LogisticRegression()
lr_ovr.fit(X_train,y_train)
#print(lr_ovr.coef_)
#print(lr_ovr.coef_.shape)
#print(lr_ovr.intercept_)
#print(lr_ovr.intercept_.shape)
lr_ovr_pred = lr_ovr.predict(X_test)
#print(confusion_matrix(y_test,lr_ovr_pred))
print(classification_report(y_test, lr_ovr_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, lr_ovr_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.450     0.360     0.400        25
                 ForwardHead      0.609     0.500     0.549        28
          LeftArmFallForward      0.569     0.696     0.626       125
 LeftAsymmetricalWeightShift      0.462     0.429     0.444        14
        LeftKneeMovesOutward      0.600     0.273     0.375        11
       LeftShoulderElevation      0.000     0.000     0.000        13
         RightArmFallForward      0.506     0.484     0.494        91
RightAsymmetricalWeightShift      0.000     0.000     0.000         3
        RightKneeMovesInward      0.000     0.000     0.000         8
       RightKneeMovesOutward      0.943     0.943     0.943        53
      RightShoulderElevation      0.541     0.688     0.606        48

                    accuracy                          0.587       419
                   macro avg      0.425     0.397     0.403       419
                we

In [25]:
lr_mn = LogisticRegression()
lr_mn.fit(X_train,y_train)
#print(lr_mn.coef_)
#print(lr_mn.coef_.shape)
#print(lr_mn.intercept_)
#print(lr_mn.intercept_.shape)
lr_mn_pred = lr_mn.predict(X_test)
#print(confusion_matrix(y_test,lr_mn_pred))
print(classification_report(y_test, lr_mn_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, lr_mn_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.450     0.360     0.400        25
                 ForwardHead      0.609     0.500     0.549        28
          LeftArmFallForward      0.569     0.696     0.626       125
 LeftAsymmetricalWeightShift      0.462     0.429     0.444        14
        LeftKneeMovesOutward      0.600     0.273     0.375        11
       LeftShoulderElevation      0.000     0.000     0.000        13
         RightArmFallForward      0.506     0.484     0.494        91
RightAsymmetricalWeightShift      0.000     0.000     0.000         3
        RightKneeMovesInward      0.000     0.000     0.000         8
       RightKneeMovesOutward      0.943     0.943     0.943        53
      RightShoulderElevation      0.541     0.688     0.606        48

                    accuracy                          0.587       419
                   macro avg      0.425     0.397     0.403       419
                we

### Naive Bayes Classification

In [26]:
gnb = GaussianNB()

gnb_pred = gnb.fit(X_train,y_train).predict(X_test)
#print(confusion_matrix(y_test, gnb_pred))
print('Accuracy:' + str(accuracy_score(y_test, gnb_pred)))

Accuracy:0.4606205250596659


In [27]:
mnb = MultinomialNB()

mnb_pred = mnb.fit(X_train,y_train).predict(X_test)
#print(confusion_matrix(y_test, mnb_pred))
print('Accuracy:' + str(accuracy_score(y_test, mnb_pred)))

Accuracy:0.4033412887828162


### LDA

In [28]:
lda = LinearDiscriminantAnalysis()
lda_model = lda.fit(X_train, y_train)
#print(lda_model.priors_)
#print(lda_model.means_)
#print(lda_model.coef_)
lda_model_pred = lda_model.predict(X_test)
#print(confusion_matrix(y_test, lda_model_pred))
print(classification_report(y_test, lda_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, lda_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.414     0.480     0.444        25
                 ForwardHead      0.607     0.607     0.607        28
          LeftArmFallForward      0.602     0.616     0.609       125
 LeftAsymmetricalWeightShift      0.500     0.571     0.533        14
        LeftKneeMovesOutward      0.333     0.364     0.348        11
       LeftShoulderElevation      0.000     0.000     0.000        13
         RightArmFallForward      0.556     0.440     0.491        91
RightAsymmetricalWeightShift      0.250     0.333     0.286         3
        RightKneeMovesInward      0.000     0.000     0.000         8
       RightKneeMovesOutward      0.980     0.906     0.941        53
      RightShoulderElevation      0.515     0.708     0.596        48

                    accuracy                          0.575       419
                   macro avg      0.432     0.457     0.441       419
                we

### K-Nearest Neighbors Clustering K = 1 2 3

In [29]:
knn = neighbors.KNeighborsClassifier(n_neighbors=1)
knn_model = knn.fit(X_train, y_train)
#print(knn_model.priors_)
#print(knn_model.means_)
#print(knn_model.coef_)
knn_model_pred = knn.predict(X_test)
#print(confusion_matrix(y_test, knn_model_pred))
print(classification_report(y_test, knn_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, knn_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.457     0.640     0.533        25
                 ForwardHead      0.353     0.214     0.267        28
          LeftArmFallForward      0.531     0.480     0.504       125
 LeftAsymmetricalWeightShift      0.600     0.429     0.500        14
        LeftKneeMovesOutward      0.200     0.182     0.190        11
       LeftShoulderElevation      0.143     0.077     0.100        13
         RightArmFallForward      0.472     0.549     0.508        91
RightAsymmetricalWeightShift      0.000     0.000     0.000         3
        RightKneeMovesInward      0.182     0.250     0.211         8
       RightKneeMovesOutward      0.891     0.925     0.907        53
      RightShoulderElevation      0.463     0.521     0.490        48

                    accuracy                          0.518       419
                   macro avg      0.390     0.388     0.383       419
                we

In [30]:
knn = neighbors.KNeighborsClassifier(n_neighbors=2)
knn_model = knn.fit(X_train, y_train)
#print(knn_model.priors_)
#print(knn_model.means_)
#print(knn_model.coef_)
knn_model_pred = knn.predict(X_test)
#print(confusion_matrix(y_test, knn_model_pred))
print(classification_report(y_test, knn_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, knn_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.370     0.680     0.479        25
                 ForwardHead      0.308     0.286     0.296        28
          LeftArmFallForward      0.497     0.624     0.553       125
 LeftAsymmetricalWeightShift      0.500     0.500     0.500        14
        LeftKneeMovesOutward      0.286     0.364     0.320        11
       LeftShoulderElevation      0.125     0.077     0.095        13
         RightArmFallForward      0.545     0.396     0.459        91
RightAsymmetricalWeightShift      0.000     0.000     0.000         3
        RightKneeMovesInward      0.286     0.250     0.267         8
       RightKneeMovesOutward      0.907     0.925     0.916        53
      RightShoulderElevation      0.444     0.250     0.320        48

                    accuracy                          0.511       419
                   macro avg      0.388     0.395     0.382       419
                we

In [31]:
knn = neighbors.KNeighborsClassifier(n_neighbors=3)
knn_model = knn.fit(X_train, y_train)
#print(knn_model.priors_)
#print(knn_model.means_)
#print(knn_model.coef_)
knn_model_pred = knn.predict(X_test)
#print(confusion_matrix(y_test, knn_model_pred))
print(classification_report(y_test, knn_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, knn_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.413     0.760     0.535        25
                 ForwardHead      0.423     0.393     0.407        28
          LeftArmFallForward      0.511     0.536     0.523       125
 LeftAsymmetricalWeightShift      0.455     0.357     0.400        14
        LeftKneeMovesOutward      0.444     0.364     0.400        11
       LeftShoulderElevation      0.000     0.000     0.000        13
         RightArmFallForward      0.456     0.396     0.424        91
RightAsymmetricalWeightShift      0.000     0.000     0.000         3
        RightKneeMovesInward      0.400     0.250     0.308         8
       RightKneeMovesOutward      0.926     0.943     0.935        53
      RightShoulderElevation      0.442     0.479     0.460        48

                    accuracy                          0.518       419
                   macro avg      0.406     0.407     0.399       419
                we

### <font color='green'>26 Predictors</font>

Use 80% with 26 predictors (all angle deviations and nasm deviations - symmetry)

In [32]:
trainingPercent = 0.8

y_train, y_test ,X_train, X_test  = train_test_split(AimoScoreWeakLinks[["WeakLinks"]], AimoScoreWeakLinks.drop(columns=["WeakLinks","AimoScore","SCORE","ID","Date","No_6_Angle_Deviation","No_7_Angle_Deviation","No_11_Angle_Deviation","No_12_Angle_Deviation","No_13_Angle_Deviation","No_2_NASM_Deviation","No_5_NASM_Deviation","No_9_NASM_Deviation","No_12_NASM_Deviation","No_14_NASM_Deviation","No_16_NASM_Deviation","No_19_NASM_Deviation","No_2_NASM_Deviation","No_1_Time_Deviation","No_2_Time_Deviation","EstimatedScore"]), train_size = trainingPercent)

### Logestic Regression

In [33]:
lr_ovr = LogisticRegression()
lr_ovr.fit(X_train,y_train)
#print(lr_ovr.coef_)
#print(lr_ovr.coef_.shape)
#print(lr_ovr.intercept_)
#print(lr_ovr.intercept_.shape)
lr_ovr_pred = lr_ovr.predict(X_test)
#print(confusion_matrix(y_test,lr_ovr_pred))
print(classification_report(y_test, lr_ovr_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, lr_ovr_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.350     0.500     0.412        14
                 ForwardHead      0.481     0.650     0.553        20
          LeftArmFallForward      0.477     0.573     0.520       124
 LeftAsymmetricalWeightShift      0.545     0.545     0.545        11
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.000     0.000     0.000        12
       LeftShoulderElevation      0.000     0.000     0.000         7
         RightArmFallForward      0.410     0.436     0.423        94
RightAsymmetricalWeightShift      0.000     0.000     0.000         7
        RightKneeMovesInward      0.000     0.000     0.000         8
       RightKneeMovesOutward      0.875     0.860     0.867        57
      RightShoulderElevation      0.554     0.484     0.517        64

                    accuracy                          0.520       419
                  

In [34]:
lr_mn = LogisticRegression()
lr_mn.fit(X_train,y_train)
#print(lr_mn.coef_)
#print(lr_mn.coef_.shape)
#print(lr_mn.intercept_)
#print(lr_mn.intercept_.shape)
lr_mn_pred = lr_mn.predict(X_test)
#print(confusion_matrix(y_test,lr_mn_pred))
print(classification_report(y_test, lr_mn_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, lr_mn_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.350     0.500     0.412        14
                 ForwardHead      0.481     0.650     0.553        20
          LeftArmFallForward      0.477     0.573     0.520       124
 LeftAsymmetricalWeightShift      0.545     0.545     0.545        11
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.000     0.000     0.000        12
       LeftShoulderElevation      0.000     0.000     0.000         7
         RightArmFallForward      0.410     0.436     0.423        94
RightAsymmetricalWeightShift      0.000     0.000     0.000         7
        RightKneeMovesInward      0.000     0.000     0.000         8
       RightKneeMovesOutward      0.875     0.860     0.867        57
      RightShoulderElevation      0.554     0.484     0.517        64

                    accuracy                          0.520       419
                  

### Naive Bayes Classification

In [35]:
gnb = GaussianNB()

gnb_pred = gnb.fit(X_train,y_train).predict(X_test)
#print(confusion_matrix(y_test, gnb_pred))
print('Accuracy:' + str(accuracy_score(y_test, gnb_pred)))

Accuracy:0.4200477326968974


In [36]:
mnb = MultinomialNB()

mnb_pred = mnb.fit(X_train,y_train).predict(X_test)
#print(confusion_matrix(y_test, mnb_pred))
print('Accuracy:' + str(accuracy_score(y_test, mnb_pred)))

Accuracy:0.38424821002386633


### LDA

In [37]:
lda = LinearDiscriminantAnalysis()
lda_model = lda.fit(X_train, y_train)
#print(lda_model.priors_)
#print(lda_model.means_)
#print(lda_model.coef_)
lda_model_pred = lda_model.predict(X_test)
#print(confusion_matrix(y_test, lda_model_pred))
print(classification_report(y_test, lda_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, lda_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.350     0.500     0.412        14
                 ForwardHead      0.444     0.600     0.511        20
          LeftArmFallForward      0.508     0.524     0.516       124
 LeftAsymmetricalWeightShift      0.667     0.727     0.696        11
               LeftHeelRises      0.000     0.000     0.000         0
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.357     0.417     0.385        12
       LeftShoulderElevation      0.222     0.286     0.250         7
         RightArmFallForward      0.457     0.447     0.452        94
RightAsymmetricalWeightShift      0.200     0.143     0.167         7
        RightKneeMovesInward      0.000     0.000     0.000         8
       RightKneeMovesOutward      0.941     0.842     0.889        57
      RightShoulderElevation      0.542     0.500     0.520        64

                  

### K-Nearest Neighbors Clustering K = 1 2 3

In [38]:
knn = neighbors.KNeighborsClassifier(n_neighbors=1)
knn_model = knn.fit(X_train, y_train)
#print(knn_model.priors_)
#print(knn_model.means_)
#print(knn_model.coef_)
knn_model_pred = knn.predict(X_test)
#print(confusion_matrix(y_test, knn_model_pred))
print(classification_report(y_test, knn_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, knn_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.389     0.500     0.438        14
                 ForwardHead      0.476     0.500     0.488        20
          LeftArmFallForward      0.465     0.427     0.445       124
 LeftAsymmetricalWeightShift      0.312     0.455     0.370        11
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.500     0.417     0.455        12
       LeftShoulderElevation      0.154     0.286     0.200         7
         RightArmFallForward      0.364     0.426     0.392        94
RightAsymmetricalWeightShift      0.333     0.143     0.200         7
        RightKneeMovesInward      0.111     0.125     0.118         8
       RightKneeMovesOutward      0.961     0.860     0.907        57
      RightShoulderElevation      0.519     0.438     0.475        64

                    accuracy                          0.480       419
                  

In [39]:
knn = neighbors.KNeighborsClassifier(n_neighbors=2)
knn_model = knn.fit(X_train, y_train)
#print(knn_model.priors_)
#print(knn_model.means_)
#print(knn_model.coef_)
knn_model_pred = knn.predict(X_test)
#print(confusion_matrix(y_test, knn_model_pred))
print(classification_report(y_test, knn_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, knn_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.281     0.643     0.391        14
                 ForwardHead      0.400     0.600     0.480        20
          LeftArmFallForward      0.436     0.573     0.495       124
 LeftAsymmetricalWeightShift      0.222     0.364     0.276        11
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.308     0.333     0.320        12
       LeftShoulderElevation      0.167     0.286     0.211         7
         RightArmFallForward      0.342     0.266     0.299        94
RightAsymmetricalWeightShift      0.500     0.143     0.222         7
        RightKneeMovesInward      0.000     0.000     0.000         8
       RightKneeMovesOutward      0.906     0.842     0.873        57
      RightShoulderElevation      0.591     0.203     0.302        64

                    accuracy                          0.451       419
                  

In [40]:
knn = neighbors.KNeighborsClassifier(n_neighbors=3)
knn_model = knn.fit(X_train, y_train)
#print(knn_model.priors_)
#print(knn_model.means_)
#print(knn_model.coef_)
knn_model_pred = knn.predict(X_test)
#print(confusion_matrix(y_test, knn_model_pred))
print(classification_report(y_test, knn_model_pred, digits = 3))
print('Accuracy:' + str(accuracy_score(y_test, knn_model_pred)))

                              precision    recall  f1-score   support

        ExcessiveForwardLean      0.312     0.714     0.435        14
                 ForwardHead      0.280     0.350     0.311        20
          LeftArmFallForward      0.441     0.540     0.486       124
 LeftAsymmetricalWeightShift      0.333     0.455     0.385        11
               LeftHeelRises      0.000     0.000     0.000         0
         LeftKneeMovesInward      0.000     0.000     0.000         1
        LeftKneeMovesOutward      0.273     0.250     0.261        12
       LeftShoulderElevation      0.143     0.143     0.143         7
         RightArmFallForward      0.469     0.404     0.434        94
RightAsymmetricalWeightShift      0.500     0.143     0.222         7
        RightKneeMovesInward      0.500     0.250     0.333         8
       RightKneeMovesOutward      0.980     0.860     0.916        57
      RightShoulderElevation      0.513     0.312     0.388        64

                  

### QDA

In [41]:
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis

qda = QuadraticDiscriminantAnalysis()
qda_model = qda.fit(X_train, y_train)
print(qda_model.priors_)
print(qda_model.means_)
#print(qda_model.coef_)
#qda_model_pred = qda_model.predict(X_test)
#print(confusion_matrix(y_test, qda_model_pred))
#print(classification_report(y_test, qda_model_pred, digits = 3))
#print('Accuracy:' + str(accuracy_score(y_test, qda_model_pred)))

[0.05137395 0.0567503  0.30227001 0.03703704 0.00119474 0.00119474
 0.02210275 0.03046595 0.22580645 0.00716846 0.02210275 0.13142174
 0.11111111]
[[0.70001001 0.77715681 0.52475727 0.66257938 0.40784425 0.83419528
  0.4453138  0.60487916 0.40784425 0.70001001 0.60487916 0.63311757
  0.57716348 0.75642566 0.93685007 0.55198915 0.69500517 0.55392991
  0.69188549 0.64635815 0.72288212 0.74125544 0.66988088 0.69988767
  0.64198169 0.5735878 ]
 [0.66197991 0.68068162 0.71039795 0.54950288 0.53209494 0.70481512
  0.43073825 0.40578922 0.53209494 0.66197991 0.40578922 0.55457223
  0.94058748 0.5511138  0.59952176 0.51686174 0.62407813 0.62505978
  0.76515895 0.56645272 0.68507136 0.69130358 0.63329054 0.70251957
  0.64829218 0.5035868 ]
 [0.56099977 0.53259428 0.5942718  0.59333815 0.53703575 0.62396445
  0.51037497 0.46801866 0.53703575 0.56099977 0.46801866 0.58003561
  0.53344809 0.67920713 0.52569703 0.51366888 0.65465177 0.62024921
  0.56607626 0.63903861 0.70821622 0.69948951 0.6376380